src/codegen.c

```diff
@@ -0,0 +1,1298 @@
+/*
+ * rv32emu is freely redistributable under the MIT License. See the file
+ * "LICENSE" for information on usage and redistribution of this file.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "cache.h"
+#include "codegen.h"
+#include "decode.h"
+#include "softfloat.h"
+
+#define SET_SIZE 1024
+
+typedef struct {
+    uint32_t table[SET_SIZE][32];
+} set_t;
+
+static inline uint32_t hash(uint32_t key)
+{
+    return (key >> 1) & (SET_SIZE - 1);
+}
+
+void set_reset(set_t *set)
+{
+    memset(set, 0, sizeof(set_t));
+}
+
+bool set_add(set_t *set, uint32_t key)
+{
+    uint32_t index = hash(key);
+    uint8_t count = 0;
+    while (set->table[index][count] != 0) {
+        if (set->table[index][count++] == key)
+            return false;
+    }
+
+    set->table[index][count] = key;
+    return true;
+}
+
+bool set_has(set_t *set, uint32_t key)
+{
+    uint32_t index = hash(key);
+    uint8_t count = 0;
+    while (set->table[index][count] != 0) {
+        if (set->table[index][count++] == key)
+            return true;
+    }
+    return false;
+}
+
+static set_t set;
+static bool insn_is_branch(uint8_t opcode)
+{
+    switch (opcode) {
+#define _(inst, can_branch) IIF(can_branch)(case rv_insn_##inst:, )
+        RISCV_INSN_LIST
+#undef _
+        return true;
+    }
+    return false;
+}
+typedef void (*gen_func_t)(riscv_t *, const rv_insn_t *, char *, uint32_t);
+static gen_func_t dispatch_table[rv_insn_qty];
+static char funcbuf[128] = {0};
+#define RVOP(inst, code)                                                  \
```

Comment:
Instead of duplicating `RVOP` and instruction executions, try to provide JIT-aware `RVOP` macro which concatenates the string for code generation.